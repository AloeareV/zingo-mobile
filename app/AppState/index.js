import Address from './Address';
import AddressBookEntry from './AddressBookEntry';
import AppStateLoaded from './AppStateLoaded';
import AppStateLoading from './AppStateLoading';
import ConnectedCompanionApp from './ConnectedCompanionApp';
import ErrorModalData from './ErrorModalData';
import InfoType from './InfoType';
import PasswordState from './PasswordState';
import ReceivePageState from './ReceivePageState';
import SendJsonToType from './SendJsonToType';
import SendPageState from './SendPageState';
import SendProgress from './SendProgress';
import ServerSelectState from './ServerSelectState';
import SettingsFileEntry from './SettingsFileEntry';
import SyncStatus from './SyncStatus';
import SyncStatusReport from './SyncStatusReport';
import ToAddr from './ToAddr';
import TotalBalance from './TotalBalance';
import Transaction from './Transaction';
import TxDetailType from './TxDetailType';
import WalletSeed from './WalletSeed';
import WalletSettings from './WalletSettings';

export {
  AppStateLoaded,
  AppStateLoading,
  Address,
  AddressBookEntry,
  ConnectedCompanionApp,
  ErrorModalData,
  InfoType,
  PasswordState,
  ReceivePageState,
  SendJsonToType,
  SendPageState,
  SendProgress,
  ServerSelectState,
  SettingsFileEntry,
  SyncStatus,
  SyncStatusReport,
  ToAddr,
  TotalBalance,
  Transaction,
  TxDetailType,
  WalletSeed,
  WalletSettings,
};
