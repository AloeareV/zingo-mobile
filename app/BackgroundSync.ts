import RPCModule from './RPCModule';
import AsyncStorage from '@react-native-async-storage/async-storage';
import NetInfo, { NetInfoStateType } from '@react-native-community/netinfo';
import { RPCSyncStatusType } from './rpc/types/RPCSyncStatusType';

const BackgroundSync = async (task_data: any) => {
  // this not interact with the server.
  const exists = await RPCModule.walletExists();

  // only if exists the wallet file make sense to do the sync.
  if (exists && exists !== 'false') {
    // only if we have connection make sense to call RPCModule.
    const networkState = await NetInfo.fetch();
    if (
      !networkState.isConnected ||
      networkState.type === NetInfoStateType.cellular ||
      (networkState.details !== null && networkState.details.isConnectionExpensive)
    ) {
      //console.log(
      //  'BS: Not started (connected: ' + networkState.isConnected,
      //  +', type: ' +
      //    networkState.type +
      //    +', expensive connection: ' +
      //    networkState.details?.isConnectionExpensive +
      //    ')',
      //);
      return;
    }
    // if the App goes to Foreground kill the interval
    const background = await AsyncStorage.getItem('@background');
    if (background === 'no') {
      //console.log('BS: Not started (going to foreground)');
      return;
    }

    let batch_num = -1;
    console.log('BS:', task_data);

    // finishEarly has two fields: wait, and done.
    // wait() returns a promise, which is resolved when
    // done() is called
    let finishEarly = manuallyResolve();

    let saver = setInterval(async () => {
      const networkStateSaver = await NetInfo.fetch();
      if (
        !networkStateSaver.isConnected ||
        networkStateSaver.type === NetInfoStateType.cellular ||
        (networkStateSaver.details !== null && networkStateSaver.details.isConnectionExpensive)
      ) {
        //console.log(
        //  'BS: Interrupted (connected: ' + networkStateSaver.isConnected,
        //  +', type: ' +
        //    networkStateSaver.type +
        //    +', expensive connection: ' +
        //    networkStateSaver.details?.isConnectionExpensive +
        //    ')',
        //);
        clearInterval(saver);
        finishEarly.done();
        return;
      }
      // if the App goes to Foreground kill the interval
      const backgroundSaver = await AsyncStorage.getItem('@background');
      if (backgroundSaver === 'no') {
        clearInterval(saver);
        //console.log('BS: Finished (going to foreground)');
        finishEarly.done();
        return;
      }

      const syncStatusStr: string = await RPCModule.execute('syncstatus', '');
      if (syncStatusStr) {
        if (syncStatusStr.toLowerCase().startsWith('error')) {
          //console.log(`BS: Error sync status ${syncStatusStr}`);
          return;
        }
      } else {
        //console.log('BS: Internal Error sync status');
        return;
      }
      // TODO: verify this JSON parse
      const ss: RPCSyncStatusType = await JSON.parse(syncStatusStr);

      //console.log('BS:', ss);
      if (ss.batch_num && ss.batch_num > -1 && batch_num !== ss.batch_num) {
        await RPCModule.doSave();
        //console.log('BS: saving...');
        // update batch_num with the new value, otherwise never change
        batch_num = ss.batch_num;
      }
    }, 5000);

    await Promise.race([RPCModule.execute('sync', ''), finishEarly.wait()]);
    clearInterval(saver);
  } else {
    console.log('BS: wallet file does not exist');
  }
  //console.log('BS: Finished (end of syncing)');
};

export default BackgroundSync;

function manuallyResolve() {
  let resolve: Function;
  // new Promise takes a function as an argument. When that function is called
  // the promise resolves with the value output by that function.
  // By passing the function out of the promise, we can call it later
  // in order to resolve the promise at will
  const promise = new Promise(fun => {
    resolve = fun;
  });

  function done() {
    resolve();
  }

  function wait() {
    return promise;
  }

  return { wait, done };
}
