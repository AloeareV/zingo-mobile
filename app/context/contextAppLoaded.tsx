import React, { ReactNode } from 'react';
import { StackScreenProps } from '@react-navigation/stack';

import {
  AppStateLoaded,
  SyncingStatusReportClass,
  TotalBalanceClass,
  ReceivePageStateClass,
  ErrorModalDataClass,
  SendProgressClass,
  WalletSettingsClass,
  TransactionType,
  InfoType,
  SyncingStatusType,
  WalletSeedType,
  DimensionsType,
  AddressClass,
  AddressBookClass,
  zecPriceType,
  BackgroundType,
  SendPageStateClass,
  ToAddrClass,
  NetInfoType,
  BackgroundErrorType,
} from '../AppState';

export const defaultAppStateLoaded: AppStateLoaded = {
  navigation: {} as StackScreenProps<any>['navigation'],
  route: {} as StackScreenProps<any>['route'],
  dimensions: {} as DimensionsType,
  appState: '' as string,
  netInfo: {} as NetInfoType,

  syncingStatusReport: new SyncingStatusReportClass(),
  totalBalance: new TotalBalanceClass(),
  addressPrivateKeys: new Map(),
  addresses: [] as AddressClass[],
  addressBook: [] as AddressBookClass[],
  transactions: [] as TransactionType[],
  sendPageState: new SendPageStateClass(new ToAddrClass(0)),
  receivePageState: new ReceivePageStateClass(''),
  info: {} as InfoType,
  walletSettings: new WalletSettingsClass(),
  syncingStatus: {} as SyncingStatusType,
  errorModalData: new ErrorModalDataClass('', ''),
  sendProgress: new SendProgressClass(0, 0, 0),
  walletSeed: {} as WalletSeedType,
  isMenuDrawerOpen: false,
  selectedMenuDrawerItem: '' as string,
  aboutModalVisible: false,
  computingModalVisible: false,
  settingsModalVisible: false,
  infoModalVisible: false,
  rescanModalVisible: false,
  seedViewModalVisible: false,
  seedChangeModalVisible: false,
  seedBackupModalVisible: false,
  seedServerModalVisible: false,
  syncReportModalVisible: false,
  poolsModalVisible: false,
  newServer: '' as string,
  uaAddress: '' as string,

  server: '' as string,
  currency: '',
  language: 'en',

  zecPrice: {
    zecPrice: 0,
    date: 0,
  } as zecPriceType,
  sendAll: false,
  background: {
    batches: 0,
    date: 0,
  } as BackgroundType,

  translate: () => '',
  backgroundError: {} as BackgroundErrorType,
};

export const ContextAppLoaded = React.createContext(defaultAppStateLoaded);

type ContextProviderProps = {
  children: ReactNode;
  value: AppStateLoaded;
};

export const ContextAppLoadedProvider = ({ children, value }: ContextProviderProps) => {
  return <ContextAppLoaded.Provider value={value}>{children}</ContextAppLoaded.Provider>;
};
