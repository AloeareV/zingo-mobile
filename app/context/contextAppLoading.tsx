import React, { ReactNode } from 'react';
import { StackScreenProps } from '@react-navigation/stack';

import {
  AppStateLoading,
  DimensionsType,
  InfoType,
  TotalBalanceClass,
  WalletSeedType,
  zecPriceType,
  backgroundType,
} from '../AppState';

export const defaultAppStateLoading: AppStateLoading = {
  navigation: {} as StackScreenProps<any>['navigation'],
  route: {} as StackScreenProps<any>['route'],
  dimensions: {} as DimensionsType,
  appState: '' as string,

  screen: 0,
  actionButtonsDisabled: false,
  walletExists: false,
  walletSeed: {} as WalletSeedType,
  totalBalance: new TotalBalanceClass(),
  info: {} as InfoType,

  server: '' as string,
  currency: '',
  language: 'en',

  zecPrice: {
    zecPrice: 0,
    date: 0,
  } as zecPriceType,
  sendAll: false,
  background: {
    batches: 0,
    date: 0,
  } as backgroundType,

  translate: () => '',
};

export const ContextAppLoading = React.createContext(defaultAppStateLoading);

type ContextProviderProps = {
  children: ReactNode;
  value: AppStateLoading;
};

export const ContextAppLoadingProvider = ({ children, value }: ContextProviderProps) => {
  return <ContextAppLoading.Provider value={value}>{children}</ContextAppLoading.Provider>;
};
