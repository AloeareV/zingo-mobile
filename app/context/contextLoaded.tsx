import React, { ReactNode } from 'react';

import {
  AppStateLoaded,
  SyncStatusReport,
  TotalBalance,
  SendPageState,
  ReceivePageState,
  ToAddr,
  ErrorModalData,
  SendProgress,
  WalletSettings,
} from '../AppState';

const defaultAppState: AppStateLoaded = {
  navigation: null,
  route: null,
  dimensions: {} as {
    width: number;
    height: number;
    orientation: 'portrait' | 'landscape';
    deviceType: 'tablet' | 'phone';
    scale: number;
  },

  syncStatusReport: new SyncStatusReport(),
  totalBalance: new TotalBalance(),
  addressPrivateKeys: new Map(),
  addresses: [],
  addressBook: [],
  transactions: null,
  sendPageState: new SendPageState(new ToAddr(0)),
  receivePageState: new ReceivePageState(),
  info: null,
  rescanning: false,
  wallet_settings: new WalletSettings(),
  syncingStatus: null,
  errorModalData: new ErrorModalData(),
  txBuildProgress: new SendProgress(),
  walletSeed: null,
  isMenuDrawerOpen: false,
  selectedMenuDrawerItem: '',
  aboutModalVisible: false,
  computingModalVisible: false,
  settingsModalVisible: false,
  infoModalVisible: false,
  rescanModalVisible: false,
  seedViewModalVisible: false,
  seedChangeModalVisible: false,
  seedBackupModalVisible: false,
  seedServerModalVisible: false,
  syncReportModalVisible: false,
  poolsModalVisible: false,
  newServer: null,
  uaAddress: null,

  translate: () => {},
  openErrorModal: () => {},
  closeErrorModal: () => {},
  toggleMenuDrawer: () => {},
  setComputingModalVisible: () => {},
  syncingStatusMoreInfoOnClick: async () => {},
  poolsMoreInfoOnClick: async () => {},
  startRescan: () => {},
};

export const ContextLoaded = React.createContext(defaultAppState);

type ContextProviderProps = {
  children: ReactNode;
  value: AppStateLoaded;
};

export const ContextLoadedProvider = ({ children, value }: ContextProviderProps) => {
  return <ContextLoaded.Provider value={value}>{children}</ContextLoaded.Provider>;
};
