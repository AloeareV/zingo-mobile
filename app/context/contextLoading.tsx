import React, { ReactNode } from 'react';
import { StackScreenProps } from '@react-navigation/stack';

import { AppStateLoading, DimensionsType, InfoType, TotalBalanceClass, WalletSeedType } from '../AppState';

export const defaultAppStateLoading: AppStateLoading = {
  navigation: {} as StackScreenProps<any>['navigation'],
  route: {} as StackScreenProps<any>['route'],
  dimensions: {} as DimensionsType,

  screen: 0,
  actionButtonsDisabled: false,
  walletExists: false,
  walletSeed: {} as WalletSeedType,
  server: '' as string,
  currency: '',
  language: 'en',
  totalBalance: new TotalBalanceClass(),
  info: {} as InfoType,

  translate: () => '',
};

export const ContextLoading = React.createContext(defaultAppStateLoading);

type ContextProviderProps = {
  children: ReactNode;
  value: AppStateLoading;
};

export const ContextLoadingProvider = ({ children, value }: ContextProviderProps) => {
  return <ContextLoading.Provider value={value}>{children}</ContextLoading.Provider>;
};
