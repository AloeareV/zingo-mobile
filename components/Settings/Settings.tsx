/* eslint-disable react-native/no-inline-styles */
import React, { useContext, useEffect, useState } from 'react';
import { View, ScrollView, SafeAreaView, Image, TouchableOpacity, TextInput } from 'react-native';
import { useTheme } from '@react-navigation/native';
import { FontAwesomeIcon } from '@fortawesome/react-native-fontawesome';
import { faDotCircle } from '@fortawesome/free-solid-svg-icons';
import { faCircle as farCircle } from '@fortawesome/free-regular-svg-icons';
import Toast from 'react-native-simple-toast';

import RegText from '../Components/RegText';
import FadeText from '../Components/FadeText';
import BoldText from '../Components/BoldText';
import ZecAmount from '../Components/ZecAmount';
import { parseServerURI, serverUris } from '../../app/uris';
import Button from '../Button';
import { ThemeType } from '../../app/types';
import { ContextLoaded } from '../../app/context';
import moment from 'moment';
import 'moment/locale/es';

type SettingsProps = {
  closeModal: () => void;
  set_wallet_option: (name: string, value: string) => void;
  set_server_option: (name: 'server' | 'currency' | 'language' | 'sendAll', value: string) => void;
  set_currency_option: (name: 'server' | 'currency' | 'language' | 'sendAll', value: string) => void;
  set_language_option: (name: 'server' | 'currency' | 'language' | 'sendAll', value: string, reset: boolean) => void;
  set_sendAll_option: (name: 'server' | 'currency' | 'language' | 'sendAll', value: boolean) => void;
};

type Memos = {
  value: string;
  text: string;
};

const Settings: React.FunctionComponent<SettingsProps> = ({
  set_wallet_option,
  set_server_option,
  set_currency_option,
  set_language_option,
  set_sendAll_option,
  closeModal,
}) => {
  const context = useContext(ContextLoaded);
  const { walletSettings, totalBalance, info, translate } = context;
  const memosArray: string = translate('settings.memos');
  let MEMOS: Memos[] = [];
  if (typeof memosArray === 'object') {
    MEMOS = memosArray as Memos[];
  }
  const { colors } = useTheme() as unknown as ThemeType;

  const [memos, setMemos] = useState(walletSettings.download_memos);
  const [filter, setFilter] = useState(walletSettings.transaction_filter_threshold);
  const [server, setServer] = useState(walletSettings.server);
  const [currency, setCurrency] = useState(walletSettings.currency);
  const [language, setLanguage] = useState(walletSettings.language);
  const [sendAll, setSendAll] = useState(walletSettings.sendAll);
  const [customIcon, setCustomIcon] = useState(farCircle);

  moment.locale(language);

  useEffect(() => {
    setCustomIcon(serverUris().find((s: string) => s === server) ? farCircle : faDotCircle);
  }, [server]);

  const currencies = (): string[] => {
    return ['', 'USD'];
  };

  const languages = (): string[] => {
    return ['en', 'es'];
  };

  const sendAlls = (): boolean[] => {
    return [true, false];
  };

  const saveSettings = async () => {
    if (
      walletSettings.download_memos === memos &&
      walletSettings.server === server &&
      walletSettings.transaction_filter_threshold === filter &&
      walletSettings.currency === currency &&
      walletSettings.language === language
    ) {
      Toast.show(translate('settings.nochanges'), Toast.LONG);
      return;
    }
    if (!memos) {
      Toast.show(translate('settings.ismemo'), Toast.LONG);
      return;
    }
    if (!filter) {
      Toast.show(translate('settings.isthreshold'), Toast.LONG);
      return;
    }
    if (!server) {
      Toast.show(translate('settings.isserver'), Toast.LONG);
      return;
    }
    if (!language) {
      Toast.show(translate('settings.islanguage'), Toast.LONG);
      return;
    }
    const result = parseServerURI(server);
    if (result.toLowerCase().startsWith('error')) {
      Toast.show(translate('settings.isuri'), Toast.LONG);
      return;
    }

    if (walletSettings.download_memos !== memos) {
      set_wallet_option('download_memos', memos);
    }
    if (walletSettings.transaction_filter_threshold !== filter) {
      set_wallet_option('transaction_filter_threshold', filter);
    }
    if (walletSettings.currency !== currency) {
      set_currency_option('currency', currency);
    }
    if (walletSettings.sendAll !== sendAll) {
      set_sendAll_option('sendAll', sendAll);
    }
    // the last one
    if (walletSettings.server !== server) {
      if (walletSettings.language !== language) {
        set_language_option('language', language, false);
      }
      set_server_option('server', server);
    } else {
      if (walletSettings.language !== language) {
        set_language_option('language', language, true);
      }
    }

    closeModal();
  };

  //console.log(walletSettings);

  return (
    <SafeAreaView
      style={{
        display: 'flex',
        justifyContent: 'flex-start',
        alignItems: 'stretch',
        height: '100%',
        backgroundColor: colors.background,
      }}>
      <View
        style={{
          display: 'flex',
          alignItems: 'center',
          paddingBottom: 10,
          backgroundColor: colors.card,
          zIndex: -1,
          paddingTop: 10,
        }}>
        <Image
          source={require('../../assets/img/logobig-zingo.png')}
          style={{ width: 80, height: 80, resizeMode: 'contain' }}
        />
        <ZecAmount
          currencyName={info.currencyName ? info.currencyName : ''}
          size={36}
          amtZec={totalBalance.total}
          style={{ opacity: 0.5 }}
        />
        <RegText color={colors.money} style={{ marginTop: 5, padding: 5 }}>
          {translate('settings.title')}
        </RegText>
        <View style={{ width: '100%', height: 1, backgroundColor: colors.primary }} />
      </View>

      <ScrollView
        style={{ maxHeight: '85%' }}
        contentContainerStyle={{
          flexDirection: 'column',
          alignItems: 'stretch',
          justifyContent: 'flex-start',
        }}>
        <View style={{ display: 'flex', margin: 10 }}>
          <BoldText>{translate('settings.sendall-title')}</BoldText>
        </View>

        <View style={{ display: 'flex', marginLeft: 25 }}>
          {sendAlls().map((curr: boolean) => (
            <TouchableOpacity
              key={'touch-' + curr.toString()}
              style={{ marginRight: 10, marginBottom: 5, maxHeight: 50, minHeight: 48 }}
              onPress={() => setSendAll(curr as boolean)}>
              <View style={{ display: 'flex', flexDirection: 'row', marginTop: 10 }}>
                <FontAwesomeIcon icon={curr === sendAll ? faDotCircle : farCircle} size={20} color={colors.border} />
                <RegText key={'tex-' + curr.toString()} style={{ marginLeft: 10 }}>
                  {translate('settings.sendall-' + curr)}
                </RegText>
              </View>
            </TouchableOpacity>
          ))}
        </View>

        <View style={{ display: 'flex', margin: 10 }}>
          <BoldText>{translate('settings.currency-title')}</BoldText>
        </View>

        <View style={{ display: 'flex', marginLeft: 25 }}>
          {currencies().map((curr: string) => (
            <TouchableOpacity
              key={'touch-' + curr}
              style={{ marginRight: 10, marginBottom: 5, maxHeight: 50, minHeight: 48 }}
              onPress={() => setCurrency(curr as 'USD' | '')}>
              <View style={{ display: 'flex', flexDirection: 'row', marginTop: 10 }}>
                <FontAwesomeIcon icon={curr === currency ? faDotCircle : farCircle} size={20} color={colors.border} />
                <RegText key={'tex-' + curr} style={{ marginLeft: 10 }}>
                  {translate('settings.currency-' + curr)}
                </RegText>
              </View>
            </TouchableOpacity>
          ))}
        </View>

        <View style={{ display: 'flex', margin: 10 }}>
          <BoldText>{translate('settings.language-title')}</BoldText>
        </View>

        <View style={{ display: 'flex', marginLeft: 25 }}>
          {languages().map((lang: string) => (
            <TouchableOpacity
              key={'touch-' + lang}
              style={{ marginRight: 10, marginBottom: 5, maxHeight: 50, minHeight: 48 }}
              onPress={() => setLanguage(lang as 'en' | 'es')}>
              <View style={{ display: 'flex', flexDirection: 'row', marginTop: 10 }}>
                <FontAwesomeIcon icon={lang === language ? faDotCircle : farCircle} size={20} color={colors.border} />
                <RegText key={'tex-' + lang} style={{ marginLeft: 10 }}>
                  {translate('settings.language-' + lang)}
                </RegText>
              </View>
            </TouchableOpacity>
          ))}
        </View>

        <View style={{ display: 'flex', margin: 10 }}>
          <BoldText>{translate('settings.server-title')}</BoldText>
        </View>

        <View style={{ display: 'flex', marginLeft: 25 }}>
          {serverUris().map((uri: string) => (
            <TouchableOpacity
              key={'touch-' + uri}
              style={{ marginRight: 10, marginBottom: 5, maxHeight: 50, minHeight: 48 }}
              onPress={() => setServer(uri)}>
              <View style={{ display: 'flex', flexDirection: 'row', marginTop: 10 }}>
                <FontAwesomeIcon icon={uri === server ? faDotCircle : farCircle} size={20} color={colors.border} />
                <RegText key={'tex-' + uri} style={{ marginLeft: 10 }}>
                  {uri}
                </RegText>
              </View>
            </TouchableOpacity>
          ))}

          <View style={{ display: 'flex', flexDirection: 'row' }}>
            <TouchableOpacity
              style={{ marginRight: 10, marginBottom: 5, maxHeight: 50, minHeight: 48 }}
              onPress={() => setServer('')}>
              <View style={{ display: 'flex', flexDirection: 'row', marginTop: 10 }}>
                {customIcon && <FontAwesomeIcon icon={customIcon} size={20} color={colors.border} />}
                <RegText style={{ marginLeft: 10 }}>{translate('settings.custom')}</RegText>
              </View>
            </TouchableOpacity>

            {customIcon === faDotCircle && (
              <View
                accessible={true}
                accessibilityLabel={translate('settings.server-acc')}
                style={{
                  borderColor: colors.border,
                  borderWidth: 1,
                  marginLeft: 5,
                  width: 'auto',
                  maxWidth: '80%',
                  maxHeight: 48,
                  minWidth: '50%',
                  minHeight: 48,
                }}>
                <TextInput
                  placeholder={'... http------.---:--- ...'}
                  placeholderTextColor={colors.placeholder}
                  style={{
                    color: colors.text,
                    fontWeight: '600',
                    fontSize: 18,
                    minWidth: '50%',
                    minHeight: 48,
                    marginLeft: 5,
                  }}
                  value={server}
                  onChangeText={(text: string) => setServer(text)}
                  editable={true}
                  maxLength={100}
                />
              </View>
            )}
          </View>
        </View>

        <View style={{ display: 'flex', margin: 10 }}>
          <BoldText>{translate('settings.threshold-title')}</BoldText>
        </View>

        <View style={{ display: 'flex', marginLeft: 25 }}>
          <View
            accessible={true}
            accessibilityLabel={translate('settings.threshold-acc')}
            style={{
              borderColor: colors.border,
              borderWidth: 1,
              marginLeft: 5,
              width: 'auto',
              maxWidth: '60%',
              maxHeight: 48,
              minWidth: '30%',
              minHeight: 48,
            }}>
            <TextInput
              placeholder={translate('settings.number')}
              placeholderTextColor={colors.placeholder}
              keyboardType="numeric"
              style={{
                color: colors.text,
                fontWeight: '600',
                fontSize: 18,
                minWidth: '30%',
                minHeight: 48,
                marginLeft: 5,
              }}
              value={filter}
              onChangeText={(text: string) => setFilter(text)}
              editable={true}
              maxLength={6}
            />
          </View>
        </View>

        <View style={{ display: 'flex', margin: 10 }}>
          <BoldText>{translate('settings.memo-title')}</BoldText>
        </View>

        <View style={{ display: 'flex', marginLeft: 25, marginBottom: 30 }}>
          {MEMOS.map(memo => (
            <View key={'view-' + memo.value}>
              <TouchableOpacity
                style={{ marginRight: 10, marginBottom: 5, maxHeight: 50, minHeight: 48 }}
                onPress={() => setMemos(memo.value)}>
                <View style={{ display: 'flex', flexDirection: 'row', alignItems: 'center', marginTop: 10 }}>
                  <FontAwesomeIcon
                    icon={memo.value === memos ? faDotCircle : farCircle}
                    size={20}
                    color={colors.border}
                  />
                  <RegText key={'text-' + memo.value} style={{ marginLeft: 10 }}>
                    {translate(`settings.value-${memo.value}`)}
                  </RegText>
                </View>
              </TouchableOpacity>
              <FadeText key={'fade-' + memo.value}>{memo.text}</FadeText>
            </View>
          ))}
        </View>
      </ScrollView>
      <View
        style={{
          flexGrow: 1,
          flexDirection: 'row',
          justifyContent: 'center',
          alignItems: 'center',
          marginVertical: 5,
        }}>
        <Button type="Primary" title={translate('settings.save')} onPress={saveSettings} />
        <Button type="Secondary" title={translate('cancel')} style={{ marginLeft: 10 }} onPress={closeModal} />
      </View>
    </SafeAreaView>
  );
};

export default Settings;
