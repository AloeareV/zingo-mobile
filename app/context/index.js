import { defaultAppStateLoaded, ContextAppLoaded, ContextAppLoadedProvider } from './contextAppLoaded';
import { defaultAppStateLoading, ContextAppLoading, ContextAppLoadingProvider } from './contextAppLoading';

export {
  defaultAppStateLoaded,
  ContextAppLoaded,
  ContextAppLoadedProvider,
  defaultAppStateLoading,
  ContextAppLoading,
  ContextAppLoadingProvider,
};
