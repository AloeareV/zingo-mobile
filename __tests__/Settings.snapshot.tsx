/**
 * @format
 */

import 'react-native';
import React from 'react';
import { StackScreenProps } from '@react-navigation/stack';

import { render } from '@testing-library/react-native';
import Settings from '../components/Settings';
import { ContextLoadedProvider } from '../app/context';

import {
  Address,
  AddressBookEntry,
  ErrorModalData,
  InfoType,
  ReceivePageState,
  SendPageState,
  SendProgress,
  SyncStatus,
  SyncStatusReport,
  ToAddr,
  TotalBalance,
  Transaction,
  WalletSeed,
  WalletSettings,
} from '../app/AppState';

jest.mock('@fortawesome/react-native-fontawesome', () => ({
  FontAwesomeIcon: '',
}));
jest.mock('react-native-localize', () => ({
  getNumberFormatSettings: () => {
    return {
      decimalSeparator: '.',
      groupingSeparator: ',',
    };
  },
}));
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');
jest.useFakeTimers();

// test suite
describe('Component Settings - test', () => {
  //snapshot test
  const state = {
    navigation: {} as StackScreenProps<any>['navigation'],
    route: {} as StackScreenProps<any>['route'],
    dimensions: {} as {
      width: number;
      height: number;
      orientation: 'portrait' | 'landscape';
      deviceType: 'tablet' | 'phone';
      scale: number;
    },

    syncStatusReport: new SyncStatusReport(),
    addressPrivateKeys: new Map(),
    addresses: [] as Address[],
    addressBook: [] as AddressBookEntry[],
    transactions: [] as Transaction[],
    sendPageState: new SendPageState(new ToAddr(0)),
    receivePageState: new ReceivePageState(),
    info: {} as InfoType,
    rescanning: false,
    syncingStatus: {} as SyncStatus,
    errorModalData: new ErrorModalData(),
    sendProgress: new SendProgress(),
    walletSeed: {} as WalletSeed,
    isMenuDrawerOpen: false,
    selectedMenuDrawerItem: '' as string,
    aboutModalVisible: false,
    computingModalVisible: false,
    settingsModalVisible: false,
    infoModalVisible: false,
    rescanModalVisible: false,
    seedViewModalVisible: false,
    seedChangeModalVisible: false,
    seedBackupModalVisible: false,
    seedServerModalVisible: false,
    syncReportModalVisible: false,
    poolsModalVisible: false,
    newServer: '' as string,
    uaAddress: '' as string,
    translate: (p: string) => {
      if (p === 'settings.memos') {
        return [
          {
            value: 'none',
            text: 'text none',
          },
          {
            value: 'wallet',
            text: 'text wallet',
          },
          {
            value: 'all',
            text: 'text all',
          },
        ];
      } else {
        return 'text translated';
      }
    },
    totalBalance: new TotalBalance(),
    wallet_settings: new WalletSettings(),
  };
  state.info.currencyName = 'ZEC';
  state.totalBalance.total = 1.12345678;
  state.wallet_settings.server = 'https://zcash.es';
  const onClose = jest.fn();
  const onSetOption = jest.fn();
  test('Settings - snapshot', () => {
    const settings = render(
      <ContextLoadedProvider value={state}>
        <Settings closeModal={onClose} set_wallet_option={onSetOption} set_server_option={onSetOption} />
      </ContextLoadedProvider>,
    );
    expect(settings.toJSON()).toMatchSnapshot();
  });
});
