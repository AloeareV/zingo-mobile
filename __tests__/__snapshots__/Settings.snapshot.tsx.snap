// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Settings - test Settings - snapshot 1`] = `
<RCTSafeAreaView
  style={
    {
      "alignItems": "stretch",
      "backgroundColor": "rgb(242, 242, 242)",
      "display": "flex",
      "height": "100%",
      "justifyContent": "flex-start",
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      {
        "alignItems": "stretch",
        "flexDirection": "column",
        "justifyContent": "flex-start",
      }
    }
    style={
      {
        "maxHeight": "85%",
      }
    }
    testID="settings.scrollView"
  >
    <View>
      <View
        style={
          {
            "display": "flex",
            "margin": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "rgb(28, 28, 30)",
              "fontSize": 18,
              "fontWeight": "bold",
              "opacity": 0.87,
            }
          }
        >
          text translated
        </Text>
      </View>
      <View
        style={
          {
            "display": "flex",
            "marginLeft": 25,
          }
        }
      />
      <View
        style={
          {
            "display": "flex",
            "margin": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "rgb(28, 28, 30)",
              "fontSize": 18,
              "fontWeight": "bold",
              "opacity": 0.87,
            }
          }
        >
          text translated
        </Text>
      </View>
      <View
        style={
          {
            "display": "flex",
            "marginLeft": 25,
          }
        }
      />
    </View>
  </RCTScrollView>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "flexGrow": 1,
        "justifyContent": "center",
        "marginVertical": 5,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "rgb(0, 122, 255)",
          "borderColor": "rgb(0, 122, 255)",
          "borderRadius": 10,
          "borderWidth": 2,
          "justifyContent": "center",
          "maxWidth": "90%",
          "minHeight": 48,
          "minWidth": "30%",
          "opacity": 1,
          "padding": 0,
          "paddingLeft": 20,
          "paddingRight": 20,
        }
      }
      testID="settings.button.save"
    >
      <View
        style={
          {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "center",
            "margin": 0,
            "minHeight": 48,
            "minWidth": 48,
            "padding": 0,
          }
        }
      >
        <Text
          style={
            {
              "color": "rgb(242, 242, 242)",
              "fontSize": 16,
              "fontWeight": "bold",
              "textAlign": "center",
              "textTransform": "uppercase",
            }
          }
        >
          text translated
        </Text>
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "rgb(242, 242, 242)",
          "borderColor": "rgb(0, 122, 255)",
          "borderRadius": 10,
          "borderWidth": 2,
          "justifyContent": "center",
          "marginLeft": 10,
          "maxWidth": "90%",
          "minHeight": 48,
          "minWidth": "30%",
          "opacity": 1,
          "padding": 0,
          "paddingLeft": 20,
          "paddingRight": 20,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "center",
            "margin": 0,
            "minHeight": 48,
            "minWidth": 48,
            "padding": 0,
          }
        }
      >
        <Text
          style={
            {
              "color": "rgb(0, 122, 255)",
              "fontSize": 16,
              "fontWeight": "bold",
              "textAlign": "center",
              "textTransform": "uppercase",
            }
          }
        >
          text translated
        </Text>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
